nomos-libp2p/src/behaviour/nat/address_mapper/protocols/pcp/client.rs

```diff
@@ -0,0 +1,281 @@
+    ProtocolMismatch,
+    #[error("Response port does not match request")]
+    PortMismatch,
+    #[error("Request timed out: {0}")]
+    Timeout(#[source] tokio::time::error::Elapsed),
+    #[error("Invalid response size: expected {expected} bytes, got {actual} bytes")]
+    InvalidSize { expected: usize, actual: usize },
+    #[error("Cannot get default gateway")]
+    CannotGetGateway,
+}
+
+pub const PCP_PORT: u16 = 5351;
+
+#[derive(Debug, Clone)]
+pub struct PcpConfig {
+    /// Default mapping lifetime
+    pub mapping_lifetime: Duration,
+    /// Initial retry delay
+    pub initial_retry: Duration,
+    /// Maximum retry delay
+    pub max_retry_delay: Duration,
+    /// Maximum number of retry attempts
+    pub max_retries: usize,
+    /// Request timeout duration
+    pub request_timeout: Duration,
+}
+
+impl Default for PcpConfig {
+    fn default() -> Self {
+        Self {
+            mapping_lifetime: Duration::from_secs(7200),
+            initial_retry: Duration::from_millis(250),
+            max_retry_delay: Duration::from_secs(1),
+            max_retries: 5,
+            request_timeout: Duration::from_secs(1),
+        }
+    }
+}
+
+/// PCP (RFC 6887) client for NAT port mapping.
+///
+/// Protocol flow:
+/// 1. ANNOUNCE: Discovers PCP server and learns external IP
+/// 2. MAP: Creates port mappings with nonce-based request/response matching
+/// 3. DELETE: Removes mappings via MAP with 0-byte lifetime
+///
+/// All requests use UDP port 5351 with exponential backoff retry.
+/// Nonce prevent replay attacks and ensure response authenticity.
+#[derive(Debug, Clone)]
+pub struct PcpClient {
+    client_addr: Ipv4Addr,
+    config: PcpConfig,
+}
+
+/// Active PCP client that has confirmed gateway support and can create port
+/// mappings.
+#[derive(Debug, Clone)]
+pub struct ActivePcpClient {
+    client_addr: Ipv4Addr,
+    gateway_ip: Ipv4Addr,
+    config: PcpConfig,
+}
+
+impl PcpClient {
+    /// Creates a new PCP client with the specified configuration.
+    pub const fn new(client_addr: Ipv4Addr, config: PcpConfig) -> Self {
+        Self {
+            client_addr,
+            config,
+        }
+    }
+
+    /// Tests if a PCP server is available by sending an ANNOUNCE request.
+    ///
+    /// Returns an `ActivePcpClient` if the server responds successfully, which
+    /// can then be used to create port mappings.
+    ///
+    /// Automatically discovers the default gateway for PCP communication.
+    pub async fn probe_available(self) -> Result<ActivePcpClient, PcpError> {
+        let gateway_ip = Self::get_default_gateway()?;
+
+        tracing::debug!("Probing PCP support at gateway {gateway_ip}");
+
+        let connection =
+            PcpConnection::connect(self.client_addr, gateway_ip, self.config.clone()).await?;
+
+        let request = PcpAnnounceRequest::new(self.client_addr);
+        let _: PcpAnnounceResponse = connection.send_request(&request).await?;
```

Comment:
Maybe add a comment specifying that since we got a valid response back from the gateway, it means it's reachable and speaks the PCP protocol. Else it might not be immediate to understand.